sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/regexpExpressions.scala

```diff
@@ -980,3 +980,38 @@ case class RegExpExtractAll(subject: Expression, regexp: Expression, idx: Expres
       newFirst: Expression, newSecond: Expression, newThird: Expression): RegExpExtractAll =
     copy(subject = newFirst, regexp = newSecond, idx = newThird)
 }
+
+// scalastyle:off line.size.limit
+@ExpressionDescription(
+  usage = """
+    _FUNC_(str, regexp) - Returns a count of the number of times that the regular expression pattern `regexp` is matched in the string `str`.
+  """,
+  arguments = """
+    Arguments:
+      * str - a string expression.
+      * regexp - a string representing a regular expression. The regex string should be a
+          Java regular expression.
+  """,
+  examples = """
+    Examples:
+      > SELECT _FUNC_('Steven Jones and Stephen Smith are the best players', 'Ste(v|ph)en');
+       2
+      > SELECT _FUNC_('abcdefghijklmnopqrstuvwxyz', '[a-z]{3}');
+       8
+  """,
+  since = "3.4.0",
+  group = "string_funcs")
+// scalastyle:on line.size.limit
+case class RegExpCount(left: Expression, right: Expression, replacement: Expression)
+  extends RuntimeReplaceable with InheritAnalysisRules {
```

Comment:
Do we need `InheritAnalysisRules`? Since the type coercion rule is very simple here: both of the 2 parameters should be string type.

Reply:
Let me try to remove it ...

Reply:
I have removed `InheritAnalysisRules` and have to re-implement the expr to:
```scala
case class RegExpCount(left: Expression, right: Expression) extends RuntimeReplaceable {

  override lazy val replacement: Expression =
    Size(RegExpExtractAll(left, right, Literal(0)), legacySizeOfNull = false)

  override def prettyName: String = "regexp_count"

  override def children: Seq[Expression] = Seq(left, right)

  override protected def withNewChildrenInternal(
      newChildren: IndexedSeq[Expression]): RegExpCount =
    copy(left = newChildren(0), right = newChildren(1))
}
```
and got the test failure on NULL handling:
```diff
 -- !query
 SELECT regexp_count(null, 'abc')
 -- !query schema
-struct<regexp_count(NULL, abc):int>
+struct<>
 -- !query output
-NULL
+java.lang.IllegalStateException
+Illegal RuntimeReplaceable: regexp_count(null, abc)
+Replacement is unresolved: size(regexp_extract_all(null, abc, 0), false)
```

Reply:
Probably, need to extend `ExpectsInputTypes`, I guess.